### ISSUE TITLE
MySQL error when deleting from dag_run table

### ISSUE THREAD

**Issue by hyderhussain** [teamclairvoyant/airflow-maintenance-dags#41]
Hi, I'm getting the following error when running this cleanup dag in my dev-airflow environment.

`[2019-04-17 18:23:43,965] {models.py:1788} ERROR - (_mysql_exceptions.OperationalError) (1093, "You can't specify target table 'dag_run' for update in FROM clause") [SQL: 'DELETE FROM dag_run WHERE dag_run.execution_date NOT IN (SELECT max(dag_run.execution_date) AS max_1 \nFROM dag_run GROUP BY dag_run.dag_id) AND dag_run.execution_date <= %s'] [parameters: (datetime.datetime(2019, 3, 18, 18, 17, 57, 359710, tzinfo=<TimezoneInfo [UTC, GMT, +00:00:00, STD]>),)]`

My understanding is that MySQL won't let you delete from the dag_run table if dag_run is used in the subquery. 

Have you encountered this issue? How did you get around it?

Thanks

**Comment by koushikajay** [comment]
Hi @hyderhussain ,

I had the same issue. I was able to fix the issue by modifying the query for the DagRun table as follows:
_**Replace the Below Query **_
` 
query = query.filter(age_check_column.notin_(session.query(func.max(age_check_column)).group_by(
dag_id)), and_(age_check_column <= max_date)`

_**Modified Query (Just a workaround  based on leveraging how MySQL optimizer works) **_
`sub_query = session.query(func.max(age_check_column)).group_by(dag_id).filter(age_check_column <= max_date).from_self()
query = query.filter(age_check_column.notin_(sub_query))`

Please let me know if this works.  

Thanks

**Comment by hyderhussain** [comment]
Thanks! That didn't work, but this did : 

`        sub_query = session.query(func.max(age_check_column)).group_by(dag_id).from_self()

        query = query.filter(age_check_column.notin_(sub_query), and_(age_check_column <= max_date))`

I tried your query, but it deleted everything except the "newest dag_run older than the max age" which is definitely not what we were going for. 

I also noticed that i'm getting foreign key constraint errors, so i had to add a block for the `TaskReschedule` table to the `DATABASE_OBJECTS` list.
